### PR TITLE
Feature/cgf 205/task queue logging

### PR DESCRIPTION
**Pull request by AndreyNegrov**
**Related to task:**  https://oat-sa.atlassian.net/browse/CGF-205
**Description:** Task queue logging

**How test**
change your LoggerService config to

```
<?php

return new oat\oatbox\log\LoggerService(array(
    'logger' => array(
        'class' => 'oat\\oatbox\\log\\logger\\TaoMonolog',
        'options' => array(
            'name' => 'tao',
            'handlers' => array(
                array(
                    'class' => 'oat\\oatbox\\log\\StreamHandler',
                    'options' => array(
                        '/var/www/html/data/tao/log/tao-nccer.log',
                        100
                    )
                )
            )
        )
    )
));
```

1. run the worker with the command:

`sudo php index.php 'oat\taoTaskQueue\scripts\tools\RunWorker' --log-file /var/www/html/data/tao/log/tao-worker.log --log-level WARNING`

2. run the web version of the project

**Expected result**
1. a file `/var/www/html/data/tao/log/tao-worker.log` should be created in which the worker's logs will be written
only those logs that correspond to the --log-level statuses will be written to this file
in our case these are only errors WARNING

2. a file `/var/www/html/data/tao/log/tao-nccer.log` contains the logs of the main project (not contain workers logs)
log level corresponds to DEBUG level Detailed debug information

### CODE REVIEW

File: common/oatbox/log/StreamHandler.php [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+     * @param bool            $bubble         Whether the messages that are handled can bubble up the stack or not
+     * @param int|null        $filePermission Optional file permissions (default (0644) are only for owner read/write)
+     * @param bool            $useLocking     Try to lock log file before doing any writes
+     *
+     * @throws \Exception                If a missing directory is not buildable
+     * @throws \InvalidArgumentException If stream is not a resource or string
+     */
+    public function __construct($defaultStream, int $defaultLevel = Logger::DEBUG, bool $bubble = true, $filePermission = null, bool $useLocking = false)
+    {
+        $stream = $this->getScriptParameter('--log-file-path') ?: $defaultStream;
```
Comment (alroniks) [inline common/oatbox/log/StreamHandler.php:58]:
```suggestion
        $stream = $this->getScriptParameter('--log-file') ?: $defaultStream;
```
Path in this case not needed, better to explain it in the description.

File: common/oatbox/log/StreamHandler.php [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+    /**
+     * @param string $parameter
+     * @return string|null
+     */
+    private function getScriptParameter(string $parameter): ?string
+    {
+        if (in_array($parameter, $_SERVER['argv'])) {
+            $indexValueParameter = array_search($parameter, $_SERVER['argv']) + 1;
+            return $_SERVER['argv'][$indexValueParameter];
```
Comment (alroniks) [inline common/oatbox/log/StreamHandler.php:71]: What if a parameter not found? It may produce Notice.

**Comment by alroniks** [comment]
`--log-level 400` doesn't look comfortable. It makes sense to improve it and use user-friendly text constants from the limited list of values, like `debug`, `error`, etc, but convert them into numeric value inside.

File: common/oatbox/log/StreamHandler.php [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    private function getLogLevelParameter(): ?string
+    {
+        $logLevelParameter = $this->getScriptParameter('--log-level');
+        if (!$logLevelParameter) {
+            return null;
+        }
+
+        $errorLevels = Logger::getLevels();
+        if (!isset($errorLevels[$logLevelParameter])) {
+            throw new \Exception('There is no such level of logging. Use DEBUG, INFO, NOTICE, WARNING, ERROR, CRITICAL, ALERT or EMERGENCY');
```
Comment (alroniks) [inline common/oatbox/log/StreamHandler.php:91]: Let's use the same `Logger::getLevels()` here to enumerate possible values for the parameter.

File: common/oatbox/log/StreamHandler.php [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+     * @param string $parameter
+     * @return string|null
+     */
+    private function getScriptParameter(string $parameter): ?string
+    {
+        if (!in_array($parameter, $_SERVER['argv'])) {
+            return null;
+        }
+
+        $indexValueParameter = array_search($parameter, $_SERVER['argv']) + 1;
```
Comment (alroniks) [inline common/oatbox/log/StreamHandler.php:74]: Redundant variable, let's move expression into the line below.

File: common/oatbox/log/StreamHandler.php [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+     * @param bool $bubble Whether the messages that are handled can bubble up the stack or not
+     * @param int|null $filePermission Optional file permissions (default (0644) are only for owner read/write)
+     * @param bool $useLocking Try to lock log file before doing any writes
+     *
+     * @throws \Exception                If a missing directory is not buildable
+     * @throws \InvalidArgumentException If stream is not a resource or string
+     */
+    public function __construct($defaultStream, int $defaultLevel = Logger::DEBUG, bool $bubble = true, $filePermission = null, bool $useLocking = false)
+    {
+        $stream = $this->getScriptParameter('--log-file') ?: $defaultStream;
```
Comment (alroniks) [inline common/oatbox/log/StreamHandler.php:59]: Not required, but I would move this string values to class constants, like:
```php
public const PARAM_LOG_FILE = '--log-file';
```

**Comment by AndreyNegrov** [comment]
> `--log-level 400` doesn't look comfortable. It makes sense to improve it and use user-friendly text constants from the limited list of values, like `debug`, `error`, etc, but convert them into numeric value inside.

implemented

File: common/oatbox/log/StreamHandler.php [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2021 (original work) Open Assessment Technologies SA (under the project TAO-PRODUCT);
+ *
+ */
+
```
Comment (alroniks) [inline common/oatbox/log/StreamHandler.php:21]: declare(strict_type = 1) is missed

File: common/oatbox/log/StreamHandler.php [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        return $_SERVER['argv'][$indexValueParameter];
+    }
+
+    /**
+     * @return string|null
+     * @throws \Exception
+     */
+    private function getLogLevelParameter(): ?string
+    {
+        $logLevelParameter = $this->getScriptParameter('--log-level');
```
Comment (alroniks) [inline common/oatbox/log/StreamHandler.php:84]: Same about constant.

**Review by alroniks — CHANGES_REQUESTED** [review]

File: common/oatbox/log/StreamHandler.php [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    private function getLogLevelParameter(): ?int
+    {
+        $logLevelParameter = $this->getScriptParameter(self::PARAM_LOG_LEVEL);
+        if (!$logLevelParameter) {
+            return null;
+        }
+
+        $errorLevels = Logger::getLevels();
+        if (!isset($errorLevels[$logLevelParameter])) {
+            throw new \Exception(sprintf('There is no such level of logging. Use %s', implode(', ', array_flip($errorLevels))));
```
Comment (alroniks) [inline common/oatbox/log/StreamHandler.php:95]:
```suggestion
            throw new \Exception(sprintf('There is not exist such log level. Please, use one of: %s', implode(', ', array_flip($errorLevels))));
```

**Review by alroniks — APPROVED** [review]
Now it looks good for me.

**Review by alroniks — APPROVED** [review]

File: common/oatbox/log/StreamHandler.php [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    private function getLogLevelParameter(): ?int
+    {
+        $logLevelParameter = $this->getScriptParameter(self::PARAM_LOG_LEVEL);
+        if (!$logLevelParameter) {
+            return null;
+        }
+
+        $errorLevels = Logger::getLevels();
+        if (!isset($errorLevels[$logLevelParameter])) {
+            throw new \Exception(sprintf('There is not exist such log level. Please, use one of: %s', implode(', ', array_flip($errorLevels))));
```
Comment (alexeyvy) [inline common/oatbox/log/StreamHandler.php:95]:
```suggestion
            throw new \Exception(sprintf('Such log level doesn't exist. Please, use one of: %s', implode(', ', array_flip($errorLevels))));
```

**Review by alexeyvy — APPROVED** [review]

**Pull request merged**